packages/server/src/models/Locker.ts

```diff
@@ -0,0 +1,106 @@
+import {ERRORS} from '../constants'
+
+export type RequestRelease = () => Promise<void> | void
+
+// The Locker interface defines methods for implementing a locking mechanism.
+// This is crucial for ensuring exclusive access to uploads and their metadata.
+// Following TUS recommendations, it's important to cancel locks from previous requests
+// to avoid holding locks for too long and to manage half-open TCP connections effectively.
+// The lock method includes a cancel callback to facilitate the cancellation of a request that previously acquired the lock.
+export interface Locker {
+  lock(id: string, cancelReq: RequestRelease): Promise<void>
+  unlock(id: string): Promise<void>
+}
+
+export interface MemoryLockerOptions {
+  acquireLockTimeout: number
+}
+
+class Lock {
+  public requestRelease?: RequestRelease
+}
+
+// MemoryLocker is an implementation of the Locker interface, maintaining locks in memory.
+// It ensures exclusive access to upload resources by managing locks for each upload.
+// The lock() method ensures that any previously held lock is released before acquiring a new one.
+// Lock acquisition attempts will timeout based on the specified 'acquireLockTimeout' duration.
+export class MemoryLocker implements Locker {
```

Comment:
Can we move this class into its own file?

Reply:
Sure thing